### ISSUE TITLE
zeodbpack is not removing unused blobs from blobstorage directory

### ISSUE THREAD

**Issue by neilferreira** [zodb/relstorage#192]
Hi all

When using zeopack the command will clean up the blobs associated to any files/images that were removed as a result of the pack.   This does not appear to be the case with zeodbpack.  

Is there any special setup/config requirements needed to achieve this?

Also given that we've been running zodbpack for months,  is there any way of cleaning up the existing blobs that no longer relate to any content?

**Comment by mauritsvanrees** [comment]
It works as far as I know, as long as you pass zodbpack a config that points to the blob storage. I have this in a project, and blobs are not getting out of hand:

```
<relstorage>
  pack-gc true
  blob-dir /Users/maurits/clients/.../var/blobstorage
  shared-blob-dir true
  <postgresql>
    dsn ...
  </postgresql>
</relstorage>
```

**Comment by neilferreira** [comment]
Thanks @mauritsvanrees - Turns out that I only specified blob-dir in my instance configuration and not in the zeodbpack configuration,  as a result the cleanup was never happening.

As it stands,  all the existing blobs won't be cleaned up and have no relationship to any content item anymore.  Is there any way that I can do a cleanup of the existing blobs that no longer have a corresponding content item?

**Comment by mauritsvanrees** [comment]
So you think that because you already packed the database without telling it where the blobs are, that the blobs that should have been removed will never be removed because the database basically does not even realise that they are still there? That *could* be, but I am not sure. The zodbpack might clean those up anyway.
You could try it in a test database, upload a couple of blobs, remove most of them, pack the database with your original config so all blobs remain, and then try it with your improved config and see what happens.
Does anyone know?

A script that indicates blobs that are definitely not used (not referenced by any transaction) would be useful, also with standard Data.fs filestorage instead of RelStorage. Does anyone know such a script?

**Comment by janjaapdriessen** [comment]
At minddistrict forked zodbbrowser where Sylvain has added [scripts for offline packing](https://github.com/minddistrict/zodbbrowser#id5).

We load the database and blobs on a local machine in order to compute the SQL queries and shell script to remove table rows and blob files and then execute these scripts on production, because running a pack in production never succeeded.

These scripts might be of use to you, and zodbbrowser is not even the right repo to put these scripts in. We're happy to move these scripts somewhere else, if anybody is interested.

**Comment by jimfulton** [comment]
On Mon, Nov 20, 2017 at 1:29 AM, Jan-Jaap Driessen <notifications@github.com
> wrote:
...

> We load the database and blobs on a local machine in order to compute the
> SQL queries and shell script to remove table rows and blob files and then
> execute these scripts on production, because running a pack in production
> never succeeded.
>
How did running pack in production fail?

**Comment by janjaapdriessen** [comment]
I'll ask Sylvain to explain more here when he is back from vacation next week. 

We're running Relstorage 1.6 on Postgres, zodbpack on our production database (112M rows in `object_state`) would time out. We're aiming to upgrade to Relstorage v2, hope we can use regular zodbpack then.

**Comment by jimfulton** [comment]
On Mon, Nov 20, 2017 at 8:38 AM, Jan-Jaap Driessen <notifications@github.com
> wrote:
...

> We're running Relstorage 1.6 on Postgres, zodbpack on our production
> database (112M rows in object_state) would time out.
>
Time out? (I'm happy to wait for more details form Sylvain. :) )

**Comment by thefunny42** [comment]
The pack did not really fail, there would be errors: some objects would be removed then they should not have been, resulting in POSKey errors. The database is in history free mode, and that was a long time ago (3~4 years). Those problems might be fixed now. In addition to this, the pack is not really efficient, and that would degrade the performances of our application while it was running, quite significantly and take way too long (more than a night).

We started to patch zodbbrowser in order to find out those POSKey errors and repair them. We built a database of references between objects in order to be able to explore the references in both directions. Additionally we can perform a sanity check with this database to ensure we do not have any POSKey errors. After, using this database we got the idea of computing an offline pack: that would be a SQL file (or multiple) with delete statements, and a shell script for the blobs. For the blobs we just convert the OID of the object to delete to an rm statement.

The advantage of this solution is that all the computation and verification can happen on a different server that is not production and does not degrade performance, and we are left just by cleaning up the data on the production server.

We have been quite happy with this solution so far, but it is not generic: our solution only works for history free databases.

To reply the question here, finding out blobs that are not used is fairly easy: make a list of all blobs on the filesystem, for each OID in your database convert it to a blob path (there's an utility function for that you can select depending of the layout), and remove it from the list if it is there. At the end you should be left with the unused blobs you can delete.

**Comment by jimfulton** [comment]
A few comments:

- POSKeyErrors after pack are worrisome, but given how widely RelStorage is used, I have to assume this is no longer an issue. @jamadden any thoughts?

- IMO, it's fine if pack took a long time.  Spreading the work out further could mitigate performance impacts, depending on your packing requirements.  At ZC, we kept 10 days of unpacked data, which was often useful for forensics.  In that sort of scenario, it would be fine to let packing take several days.

- Of course, packing  using replicated data is certainly an opportunity. That's what I did with zc.zodbdgc and ZEO.
